Reply notification

pellenilsson opened the pull request:
This PR introduces possibility for posters to get email notifications when people reply to their posts. It is based on work done by @tsileo in 2013(!), and additions I made this year. The original PR was #50.

When posting a message, there is a checkbox labelled "Subscribe to email notification of replies". Checking that checkbox will set the `notification` flag in the database for that post. Any replies made to the post will cause an email to be sent to the author of the first post.

If you click the checkbox when making a **reply**, you will get notifications for all further replies to the same parent post, so you are basically subscribing to updates to that conversation thread.

Each email will contain an unsubscribe link for **opt-out**. Clicking that will unsubscribe you from notifications for that conversation thread.

This PR also contains a fix for #395 which was triggered (or worsened) by this functionality.

This PR is however broken for many setups because of #420, so a fix for that issue is crucial.

pellenilsson commented:
The `notification` column is automatically added the database if needed when starting isso (in the constructor of `isso.db.comments`).

pellenilsson commented on isso/views/comments.py:
```diff
@@ -93,28 +97,29 @@ def dec(self, env, req, *args, **kwargs):
 
     # comment fields, that can be submitted
-    ACCEPT = set(['text', 'author', 'website', 'email', 'parent', 'title'])
+    ACCEPT = set(['text', 'author', 'website', 'email', 'parent',  'title', 'notification'])
```
Fixed.

JarJak commented:
Any news on this?

pellenilsson commented:
I consider it ready for merging. @blatinier ?

blatinier commented:
I'm soooo sorry, I miss time to test it :(
But I'm presonnally interested in it and hope to be able to do it during my holydays (coming next week)

craigphicks commented:
I merged this pull 443 and I'm getting an db error message:

sqlite3.IntegrityError: NOT NULL constraint failed: comments.notification

I suppose it is necessary to some db upgrade first.  Are there some instructions for that?

pellenilsson commented:
@craigphicks The database is supposed to be upgraded automatically when Isso is started, and that works for me. Maybe it depends on version of Python or some library. Could you try [this branch](https://github.com/pellenilsson/isso/tree/reply_notification_dbfix) where I changed the order of the database commands?

craigphicks commented:
I fetched your branch

    git fetch git@github.com:pellenilsson/isso.git reply_notification_dbfix:reno-dbfix
    git checkout reno-dbfix

Backed up the old database

    cp /var/lib/isso/comments.db /var/lib/isso/comments.pre-update.db

Ran the branch code 

    python -m pdb ./bin/isso -c /etc/isso.cfg run

Unlike with the *reply_notification* branch,  with the *reply_notification_dbfix* there is no error on initial display of messages.

However, when adding a *new* comment *isso* returns 500 and the new comment is not displayed. The pre-exisitng  messages, which are presumably read from the successfully updated db, are displayed normally.

Checking the *isso* log file shows this diagnostic information:


connected to SMTP server
connected to https://example.net/
POST /new
Traceback (most recent call last):
  File "/home/user/work/isso-positiv/isso/__init__.py", line 138, in dispatch
    response = handler(request.environ, request, **values)
  File "/home/user/work/isso-positiv/isso/views/comments.py", line 92, in dec
    return func(self, env, req, *args, **kwargs)
  File "/home/user/work/isso-positiv/isso/views/__init__.py", line 46, in dec
    return func(cls, env, req, *args, **kwargs)
  File "/home/user/work/isso-positiv/isso/views/comments.py", line 296, in new
    rv = self.comments.add(uri, data)
  File "/home/user/work/isso-positiv/isso/db/comments.py", line 68, in add
    uri)
  File "/home/user/work/isso-positiv/isso/db/__init__.py", line 62, in execute
    return con.execute(sql, args)
sqlite3.IntegrityError: NOT NULL constraint failed: comments.notification
< repeated for every new message attempt>

Restarting several times and the behavior is always the same.

craigphicks commented:
After updating the js files correctly, function seems good (*1).  It is a major improvement.  (*1 - so far I've only correctly tested with the *reply_notification_dbfix* branch, not *443*)



1. The mail response to the original poster occurs before the moderator has approved the post.  This leaves an opening for unverified anonymous  spammers to use an isso installation as a broadcast base.  In the worst case, even if guard is enabled, a spammer with multiple addresses could cause an annoying number of spam mails to be sent to a single mailbox. 

2. The original poster, having themselves been moderated, may be expecting the reply also to be moderated.  There's nothing next to the notify check box to indicate otherwise.  If advertising mail suddenly appears in a posters mailbox, they may claim an infraction of GDPR regulations, but the isso site owner wouldn't even get the joy of collecting advertising royalties.  

3. Supposing the program is changed so that replies are moderated before mail is sent to the poster.  Still, the site administrator policy of enabling or disabling moderation now has different implications than it did before reply-notification.  When disabling, whereas before it would only be the site itself upon which spam could appear, it now also includes the possibility of spam appearing in a posters personal mailbox.

craigphicks commented:
I've updated [pull 458](https://github.com/posativ/isso/pull/458), to conform to your system of prioritizing the config file *general/public_endpoint* setting.  

It is worth adding that there already exists a config setting *general/name* which may have overlapping functionality with config setting *general/public_endpoint*.  

When setting up a second isso, and serving from an nginx location 'isso-2': 

```
location /isso-2 {
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Script-Name /isso-2;      // <---- critical for subdirectory config
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_pass http://localhost:1234;
}

```
I found it was necessary to set 

```
[general]
name = isso-2
```

in order to get the connection to work.  Indeed the *isso* documentation on [Multiple Sites](https://posativ.org/isso/docs/setup/multiple-sites/) explains that it is necessary to specify the subdirectory in config under *general/name*.

pellenilsson commented:
Regarding your comments 1 and 2, I totally agree that notifications should only be sent after a comments gets through moderation. Apart from the spam issue, it would just be confusing to get a notification about a comment which is not yet visible on the site.

Regarding comment 3, you are basically saying that it's unwise to enable notifications while moderation is disabled. Do you think it should even be forbidden, or would it be enough to advice against it in the docs, and maybe print a warning at startup of Isso?

I noticed two more shortcomings:

1. Setting `notify=smtp` now enables both notifications to site administrator and to posters, there is no way to enable/disable them independently.

2. If you *don't* set `notify=smtp`, the checkbox to receive notifications is still displayed but has no effect.

craigphicks commented:
That's a pretty complex and tough topic.  My instinct is to allow all functionality, to make the information readily available, and to make it difficult to accidentally make a mistake.  The related parameters could live in the database and be enabled through the admin GUI in a coordinated way where an explanation would have to be read.  But that a lot more work and I trust your judgment.

hamoid commented:
I find this issue very important to enable conversations in our websites. Thanks to all of you who are improving isso because that helps us switch away from alternatives that track users :+1:

pellenilsson commented:
I now changed the behavior so that notifications are sent after moderation, if applicable. I also added new configuration option `reply-notifications` on both server and client side, and documentation (which was missing completely) with a warning about enabling notifications without moderation.

I didn't have time to look at the possible overlap between `name` and `public-endpoint`. I will now be mostly offline for a few weeks.

blatinier commented:
It works great here! Thanks for your great work!

pozitron57 commented:
Thank all of you who worked on this. It was the only thing I missed in Isso. Just tested — it works! As far as I understand, it is possible to subscribe to replies to your comment but not to all comments in the thread?
